content/blog/2019-10-18-double-checked-locking/index.md

```diff
@@ -0,0 +1,353 @@
+        Helper h = helper;
+        if (h == null) {
+            synchronized(this) {    
+                h = helper;
+                if (h == null) {
+                    h = new Helper();
+                    helper = h;
+                }
+            }
+        }
+        return h;
+    }
+}
+
+```
+
+In cases that the `helper` in already initialized, 
+   this optimization can reduce one volatile read by returning the local variable.
+
+### Explicit Memory Barrier
+
+The last fix with two synchronized sections does not work because releasing lock
+is an implicit "one-way" memory barrier.
+It is possible to make the double-checked locking actually work with an
+explicit memory barrier.
+For example, in C++11 we can safely implement double-checked locking with 
+`std::atomic` and `std::atomic_thread_fence`.
+
+```C++
+class Foo {
+    private:
+        std::atomic <Foo*> helper;
+    public:
+        Foo* get_helper() {
+            Foo* h = helper.load(std::memory_order_relaxed);
+            std::atomic_thread_fence(std::memory_order_acquire);        //memory barrier
+            if (h == nullptr) {
+                std::lock_guard<std::mutex> lock(m_init);
+                h = helper.load(std::memory_order_relaxed);
+                if (h == nullptr) {
+                    h = new Helper;
+                    std::atomic_thread_fence(std::memory_order_release);//memory barrier
+                    helper.store(h, std::memory_order_relaxed);
+                }
+            }
+            return tmp;
+        }
+};
+```
+
+* `std::atomic_thread_fence(std::memory_order_acquire)` guarantees that 
+read/write after a memory barrier cannot be reordered before the memory barrier.
+* `std::atomic_thread_fence(std::memory_order_release)` guarantees that 
+read/write before a memory barrier cannot be reordered after the memory barrier.
+
+Similar to the Java volatile case, 
+the initialization must be done before it stores `h` to `helper` 
+with the memory barrier.
+
+### Static Singleton
+
+If the `helper` is static, i.e. all the instances of class `Foo` shares the 
+same instance of `helper`, defining the `helper` in a static field of a separate
+class will solve the problem.
+
+```Java
+class Foo {
+    private static class HelperSingleton {
+        public static final Helper helper = new Helper();
+    }
+
+    public Helper getHelper() {
+        return HelperSingleton.helper;
+    }
+}
+```
+
+This is known as **initialization-on-demand holder idiom**, 
+which is a safe and efficient concurrent lazy initialization 
+for all Java versions.
+
+* **Lazy Initialization**: as specified by the Java language,
+    the static class `HelperSingleton` is not initialized until the first it is
+    referenced.
+* **Safe**: the class initialization is guaranteed to be synchronized by 
+    Java language.
+* **Efficient**: all subsequent calls of `getHelper` will return the initialized
+    `helper` without any synchronization overhead.
```

Comment:
Wow—I'm not sure I understand this deeply Java-specific trick. You say that class initialization is guaranteed to be "synchronized," but synchronized _with what_?

That is, regardless of how it goes about it, a safe version of this will need to somehow use hardware synchronization to order the initialization with the use. So it does not suffice just to have all initializations be synchronized _with each other_. How does the JVM avoid acquiring a lock on subsequent accesses to `HelperSingleton.helper`?